File: subsetsii.js
//Objective is the same as 'Subsets' but now we have duplicate elements.
//We also cannot have repeating sets.

let nums = [1,2,2]


//O(N!/(N - k)! * k!) solution (same as 'Subsets')

//We sort the elements first to makes sure we skip any repeating elements
nums.sort((a,b) => {
    return a - b
})

let temp = []
let index = 0
let result = []

backtrack(temp, index)

function backtrack(temp, index) {
    result.push([...temp])

    for (let i = index; i < nums.length; i++) {
        //skip any of the same elements
        if (i !== index && nums[i] == nums[i - 1]) {
            continue
        }

        backtrack([...temp, nums[i]], i + 1)
    }
}

return result

File: subsetsii.java
//Java Solution

class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        LinkedList<Integer> temp = new LinkedList<>();
        int index = 0;
        
        backtrack(temp, result, index, nums);
        return result;
    }
    
    public void backtrack(LinkedList<Integer> temp, List<List<Integer>> result, int index, int[] nums) {
        result.add(new ArrayList<>(temp));
        
        for (int i = index; i < nums.length; i++) {
            if (i != index && nums[i] == nums[i - 1]) {
                continue;
            }
            temp.add(nums[i]);
            backtrack(temp, result, i + 1, nums);
            temp.removeLast();
        }
    }
}